python-package/lightgbm/basic.py

```diff
@@ -3503,7 +3503,7 @@ def predict(self, data, start_iteration=0, num_iteration=None,
                                  raw_score, pred_leaf, pred_contrib,
                                  data_has_header, is_reshape)
 
-    def refit(self, data, label, decay_rate=0.9, **kwargs):
+    def refit(self, data, label, decay_rate=0.9, kwargs_for_predict=None, kwargs_for_dataset=None):
```

Comment:
The proposed approach is a breaking change. It means that code which previously passed any of the [prediction parameters](https://lightgbm.readthedocs.io/en/latest/Parameters.html#predict-parameters) would now have to be changed to bundle those up into a dictionary passed to `kwargs_for_predict`.

Will you consider this suggestion? It would allow the behavior that you want (fully customize the inner call to `Dataset()`) without breaking any users' existing code.

```suggestion
    def refit(
        self,
        data,
        label,
        decay_rate=0.9,
        reference=None,
        weight=None,
        group=None,
        init_score=None,
        feature_name='auto',
        categorical_feature='auto',
        dataset_params=None,
        free_raw_data=True,
        **kwargs
     ):
```

Specifically, I'm suggesting just exposing all of the keyword arguments from the `Dataset` construct ([link](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Dataset.html#lightgbm.Dataset)) and then passing them directly into it.

That would avoid breaking existing code and would make documentation and auto-completion for people working in IDEs.

@StrikerRUS @jmoralez what do you think?

Reply:
TBH, I'm OK with any of these two interfaces. Only one thing that I'm against of is param names. I believe `kwargs_for_dataset` and `kwargs_for_predict` should be `params_for_dataset` and `params_for_predict ` respectively.

Reply:
I think of `params` as anything that shows up in https://lightgbm.readthedocs.io/en/latest/Parameters.html and is an attribute of the `config` object on the C++ side.

That description doesn't apply to things like  `init_score`, `weight`, and `group`.

That's why I've proposed `dataset_params` as a keyword argument, to capture anything in https://lightgbm.readthedocs.io/en/latest/Parameters.html#dataset-parameters.

Reply:
OK. My intention was just to get rid of `kwargs` word in argument names actually. 🙂

Reply:
OK, agree on adding a single `dataset_params` argument that includes all other parameters to be passed to [Dataset](https://github.com/microsoft/LightGBM/blob/44a9120197fbfc7f3bfb9349e4eb2e9443c1971b/python-package/lightgbm/basic.py#L740) and leaving `kwargs` as it is now.

Reply:
Since @StrikerRUS didn't have a strong preference between the interfaces, please use the one I proposed above with dedicated keyword arguments for each keyword argument of `lgb.Dataset()`, and using `dataset_params` to be passed to `params` in `lgb.Dataset()`.

I feel strongly that the term "params" has a special meaning in LightGBM and including things like `init_score`, `weight` in anything called `params` could cause confusion for users.

Reply:
@jameslamb shouldn't it be all parameters from `Dataset` except `params` as `train_set = Dataset(data, label, params=new_params)` and `new_params` is defined inside the function

Reply:
I do see that at

https://github.com/microsoft/LightGBM/blob/e4c0ca5f5d428a2497944e83c4db1efae6949b04/python-package/lightgbm/basic.py#L3537-L3542

But if the goal of this PR is to support fully customizing the construction of the `Dataset` inside `Booster.refit()`, then I think it's important to include `params`.

I think it could work like this:

```python
new_params = _choose_param_value( 
     main_param_name="linear_tree", 
     params=self.params, 
     default_value=None 
 ) 
if dataset_params is None:
    dataset_params = {}

new_params.update(dataset_params)
```